### PR TITLE
ReScript v11+: 1.0.0

### PR DESCRIPTION
**Pull request by zth**
This adapts Core to better fit ReScript v11+. 

Going forward, `main` will hold Core versions `1.0.0+`, and be fully adapted to v11.
https://github.com/rescript-association/rescript-core/tree/0.x will hold `0.x` versions, and they will be compatible with `<11` as well.

### CODE REVIEW

**Comment by zth** [comment]
cc @glennsl

File: src/Core__JSON.mjs [hunk 1]
```diff
@@ -1,35 +1,34 @@
...
 
-import * as Caml_option from "rescript/lib/es6/caml_option.js";
 
 function classify(value) {
   var match = Object.prototype.toString.call(value);
   switch (match) {
     case "[object Array]" :
         return {
-                TAG: /* Array */4,
+                TAG: "Array",
```
[the remark this answers is not in the document]
Comment (zth) [reply src/Core__JSON.mjs:9]: No, this was just the first time Core had v11+ as dep.

[the remark this answers is not in the document]
Comment (cristianoc) [reply src/Core__JSON.mjs:9]: Oh right v10 is older than I'd remember.

**Review by cknitt — APPROVED** [review]
🎉

File: src/Core__List.mjs [hunk 2]
```diff
@@ -1039,7 +1072,9 @@ function equal(_l1, _l2, p) {
 function some2(l1, l2, p) {
   var _l1 = l1;
   var _l2 = l2;
-  var p$1 = Curry.__2(p);
+  var p$1 = function (a, b) {
+    return p(a, b);
+  };
```
Comment (glennsl) [inline src/Core__List.mjs:1077]: What's with the unnecessary wrapper functions here?

File: src/Core__Promise.res [hunk 3]
```diff
@@ -1,15 +1,15 @@
 type t<+'a> = promise<'a>
 
 @new
-external make: ((@uncurry (. 'a) => unit, (. 'e) => unit) => unit) => t<'a> = "Promise"
+external make: ((@uncurry 'a => unit, 'e => unit) => unit) => t<'a> = "Promise"
```
Comment (glennsl) [inline src/Core__Promise.res:4]: `@uncurry` annotations can be removed?

Comment (zth) [reply src/Core__Promise.res:4]: @glennsl I'll go ahead and just remove the `@uncurry` annotations, and we can let 1.0.0+ of Core be focused on uncurried mode (<1.0.0 can be used with curried mode in v11+ still). It'll be uncurried once it's in the compiler anyway, so seems like the best path forward.

**Pull request merged**